# lesson 1:

数据：/zfssz3/ST_DIVERSITY/UNKNOWN_Project/liqiye/PMO/PC_PA_UN/exercise/data/test.fq.gz

这是Illumina测序数据的一个例子，以fastq格式存储。  

任务是：1）网上查找资料了解fastq格式；2）统计一下这批数据产量（即总共产生了多少reads和多少碱基）；3）统计一下这批数据中A、T、C、G的总个数分别是多少。

Fastq format:  ＃学习一下fastq格式，这是测序数据最常用的存储格式。
http://en.wikipedia.org/wiki/FASTQ_format


## answer

FASTQ 格式：Fastq格式是基于文本文件的，一般用于存储生物学序列（主要是核酸序列）和对应的测序质量，核酸序列为ATCGN，质量序列是单个的ASCII字符组成。

**Fastq Example**

![fastq_example](./image/fastq_eg.JPG)


一条序列内容包括四行：
- 第一行一般是测序时的坐标信息，@开头。
- 第二行为序列信息，用ATCGN表示，N为无法判断的碱基。
- 第三行以+开始，可以添加附加信息，一般为空。
- 第四行为质量信息，由ASCII经过换算而来，即Phred值，有不同的转换算法。

## Phred值

Phred值是评估对应碱基测序质量的值，测序仪通过判断荧光信号的颜色来判断碱基的种类，ATCG分别对应不同颜色信号，信号强弱不同，在这种情况下对每个结果的判断的正确性都存在一个概率值，这个值被储存为ASCII码形式，转化方式如下：

*将该碱基判断错误概率值P取log10之后再乘以-10，得到的结果为Q。*

比如，P=1%，那么对应的Q=-10\*log10（0.01）=20（这个计算公式illumina平台使用，Solexa系列测序仪使用不同的公示来计算质量值：Q=-10log(P/1-P)）

把这个Q加上33或者64转成一个新的数值，称为Phred，最后把Phred对应的ASCII字符对应到这个碱基。

如Q=20，Phred = 20 + 33 = 53，53在ASCII码表里对应的ASCII符号是”5”。


## 数据统计

In [1]:
# lesson1 practice
# date: 2018/08/01
# author: zxzhu

filename = './data/Traning/Lesson1/test.fq'
f = open(filename, 'r')
line_num = 0
reads_num = 0
bases_num = 0
A = 0; G = 0; C = 0; T = 0; N = 0
for line in f:
    line_num += 1 
    line = line.strip()
    if line.startswith('@'):
        reads_num += 1
    if line_num % 4 == 2:
        bases_num += len(line)
        A += line.count('A')
        G += line.count('G')
        C += line.count('C')
        T += line.count('T')
        N += line.count('N')   

f.close()
print('The number of reads:{}.\nThe number of bases:{}.\n'.format(reads_num, bases_num))
print('A:{}\t G:{}\nC:{}\tT:{}\nN:{}\nDone!'.format(A,G,C,T,N))

The number of reads:2500.
The number of bases:110000.

A:33893	 G:23066
C:0	T:52291
N:750
Done!


利用Python脚本统计出：  
- reads数量：2500。
- 碱基数量：110000。
- A：33839    
- G：23066    
- C：0 
- T：52291  
- N：750

## 总结
这个lesson可以学习和掌握：
1. FASTQ文件格式
2. 简单的序列信息统计

需要深入了解：Phred值的转化和一些测序历史。

## Reference

[FASTQ和FASTA格式](https://zhuanlan.zhihu.com/p/34518389)  

[FASTQ format](https://en.wikipedia.org/wiki/FASTQ_format)

[Fastq 格式说明 & (Phred33 or Phred64)](https://blog.csdn.net/huyongfeijoe/article/details/51613827)
